e2e/testcases/helm_sync_test.go

```diff
@@ -380,3 +409,75 @@ func helmChartVersion(chartVersion string) nomostest.Sha1Func {
 		return chartVersion, nil
 	}
 }
+
+// pushes a new helm chart for use during an e2e test
+func pushHelmChart(nt *nomostest.NT, t *testing.T, helmchart, version string) string {
+	t.Helper()
+
+	nt.T.Log("Push helm chart to the artifact registry")
+	tmpDir := t.TempDir()
+
+	chartName := fmt.Sprintf("%s-%s-%s-%s", helmchart, nomostesting.GCPProjectIDFromEnv, nomostesting.GCPClusterFromEnv, generateRandomString())
+	if len(chartName) > 50 {
+		// the chartName + releaseName is used as the metadata.name of resources in the coredns helm chart, so we must trim this down
+		// to keep it under the k8s length limit
+		chartName = chartName[len(chartName)-50:]
+		chartName = strings.Trim(chartName, "-")
+	}
+
+	paths, err := filepath.Glob(fmt.Sprintf("../testdata/helm-charts/%s/*", helmchart))
+	if err != nil {
+		nt.T.Fatalf("failed to get chart filepaths: %v", err)
+	}
+	for _, path := range paths {
+		if _, err := nt.Shell.Command("cp", "-r", path, tmpDir).CombinedOutput(); err != nil {
+			nt.T.Fatalf("failed to copy helm chart: %v", err)
+		}
+	}
+	if _, err := nt.Shell.Command("sed", "-i", "-e", fmt.Sprintf("s/name: %s/name: %s/g", helmchart, chartName), filepath.Join(tmpDir, "Chart.yaml")).CombinedOutput(); err != nil {
+		nt.T.Fatalf("failed to rename helm chart: %v", err)
+	}
+	if _, err := nt.Shell.Helm("package", tmpDir, "--destination", tmpDir); err != nil {
+		nt.T.Fatalf("failed to package helm chart: %v", err)
+	}
+
+	authCmd := nt.Shell.Command("gcloud", "auth", "print-access-token")
+	loginCmd := nt.Shell.Command("helm", "registry", "login", "-uoauth2accesstoken", "--password-stdin", "https://us-docker.pkg.dev")
+	loginCmd.Stdin, err = authCmd.StdoutPipe()
+	if err != nil {
+		nt.T.Fatal("failed to setup command pipe: %v", err)
+	}
+	if err := loginCmd.Start(); err != nil {
+		nt.T.Fatalf("failed to start login command: %v", err)
+	}
+	if err := authCmd.Run(); err != nil {
+		nt.T.Fatalf("failed to run auth command: %v", err)
+	}
+	if err := loginCmd.Wait(); err != nil {
+		nt.T.Fatalf("failed to wait for login command: %v", err)
+	}
+
+	if out, err := nt.Shell.Helm("push", filepath.Join(tmpDir, fmt.Sprintf("%s-%s.tgz", chartName, version)), privateARHelmRegistry); err != nil {
+		nt.T.Fatalf("failed to push helm chart: %s; %v", string(out), err)
+	}
+	return chartName
+}
+
+// removes helm charts created during e2e testing
+func cleanHelmImages(nt *nomostest.NT, chartName string) {
+	if _, err := nt.Shell.Command("gcloud", "artifacts", "docker", "images", "delete", fmt.Sprintf("us-docker.pkg.dev/%s/config-sync-test-ar-helm/%s", nomostesting.GCPProjectIDFromEnv, chartName), "--delete-tags").CombinedOutput(); err != nil {
+		nt.T.Errorf("failed to cleanup helm chart image from registry: %v", err)
+	}
+}
+
+// generates a 5 character long random string to reduce chance of name conflict if multiple tests are running in the
+// same cluster
+func generateRandomString() string {
```

Comment:
I think something like `BUILD_ID` (prow env var) might be a better fit as a unique string that still has some meaning. Alternatively a timestamp might also provide context at least around when it was created

Reply:
Using BUILD_ID sounds good to me. Do you know how many characters long it is and if we set a unique BUILD_ID for each job? (Want to make sure that if we use BUILD_ID, that the truncation that we do later still makes sense.)

Reply:
`BUILD_ID` is unique and provided by prow: https://docs.prow.k8s.io/docs/jobs/#job-environment-variables

e.g. for your presubmit job you can see the BUILD_ID is `1664393314247380992`: https://oss.gprow.dev/view/gs/oss-prow-build-kpt-config-sync/pr-logs/pull/GoogleContainerTools_kpt-config-sync/636/kpt-config-sync-presubmit-e2e-multi-repo-test-group1/1664393314247380992

If we were to use this, we might need to double check that the env var makes it to the test process (e.g. through the [layers of test scaffolding](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/Makefile.e2e.ci#L18)).

I was thinking that using something like BUILD_ID or timestamp might provide a bit more context (e.g. find the job that created the registry). Not a huge deal though if this makes it too complicated

Reply:
I like your idea of using the prow BUILD_ID, but if we are not sure that it gets propagated to the test process then I think it is better to use a timestamp - I agree that the extra context is beneficial over a random string. I'll make that update, thanks!

Reply:
Updated to use a timestamp